drivers/led/led_pwm.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT pwm_leds
+
+/**
+ * @file
+ * @brief PWM driven LEDs
+ */
+
+#include <drivers/led.h>
+#include <drivers/pwm.h>
+#include <device.h>
+#include <zephyr.h>
+#include <sys/math_extras.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(led_pwm);
+
+#define DEV_CFG(dev)	((const struct led_pwm_config *) ((dev)->config))
+#define DEV_DATA(dev)	((struct led_pwm_data *) ((dev)->data))
+
+struct led_pwm {
+	char *pwm_label;
+	uint32_t channel;
+	uint32_t period;
+	pwm_flags_t flags;
+};
+
+struct led_pwm_config {
+	int num_leds;
+	const struct led_pwm *led;
+};
+
+struct led_pwm_data {
+	const struct device *pwm;
+};
+
+static int led_pwm_blink(const struct device *dev, uint32_t led,
+			 uint32_t delay_on, uint32_t delay_off)
+{
+	const struct led_pwm_config *config = DEV_CFG(dev);
+	struct led_pwm_data *data = DEV_DATA(dev);
+	const struct led_pwm *led_pwm;
+	uint32_t period_usec, pulse_usec;
+
+	if (led >= config->num_leds) {
+		return -EINVAL;
+	}
+
+	if (u32_add_overflow(delay_on, delay_off, &period_usec) ||
+	    u32_mul_overflow(period_usec, 1000, &period_usec) ||
+	    u32_mul_overflow(delay_on, 1000, &pulse_usec)) {
+		return -EINVAL;
+	}
+
+	led_pwm = &config->led[led];
+
+	return pwm_pin_set_usec(data[led].pwm, led_pwm->channel,
+				period_usec, pulse_usec, led_pwm->flags);
+}
+
+static int led_pwm_set_brightness(const struct device *dev,
+				  uint32_t led, uint8_t value)
+{
+	const struct led_pwm_config *config = DEV_CFG(dev);
+	struct led_pwm_data *data = DEV_DATA(dev);
+	const struct led_pwm *led_pwm;
+	uint32_t pulse;
+
+	if (led >= config->num_leds || value > 100) {
+		return -EINVAL;
+	}
+
+	led_pwm = &config->led[led];
+
+	pulse = led_pwm->period * value / 100;
+
+	return pwm_pin_set_cycles(data[led].pwm, led_pwm->channel,
+				  led_pwm->period, pulse, led_pwm->flags);
+}
+
+static int led_pwm_on(const struct device *dev, uint32_t led)
+{
+	return led_pwm_set_brightness(dev, led, 100);
+}
+
+static int led_pwm_off(const struct device *dev, uint32_t led)
+{
+	return led_pwm_set_brightness(dev, led, 0);
+}
+
+static int led_pwm_init(const struct device *dev)
+{
+	const struct led_pwm_config *config = DEV_CFG(dev);
+	struct led_pwm_data *data = DEV_DATA(dev);
+	int i;
+
+	if (!config->num_leds) {
```

Comment:
Is this check needed?

Reply:
> Is this check needed?

Maybe the driver's Kconfig dependencies can be `depends on PWM && $(dt_compat_enabled,pwm-leds)` to make it impossible to build if this is 0? Not sure what the right thing is.

If you want to keep this check, can you add a LOG_ERR() on the failure path?

One of the most common FAQs is "why is device_get_binding() returning NULL", and I want to start pushing for all driver initialization functions to log an error on failures so users can turn on logging to figure out what went wrong.

Reply:
Yes it is. This prevents the creation of the device in the case of the DT "pwm-leds" node has no child nodes.

Reply:
I added a `LOG_ERR()` to describe this error case.